nupic/research/frameworks/pytorch/hooks/track_hidden_activations.py

```diff
@@ -0,0 +1,60 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2021, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import torch
+
+from .base import TrackStatsHookBase
+
+
+class TrackHiddenActivationsHook(TrackStatsHookBase):
+    """
+    Forward hook class for tracking hidden activations of modules.
+    """
+
+    def __init__(self, name, max_samples_to_track):
+        super().__init__(name=name)
+
+        self.num_samples = max_samples_to_track
+
+        # `_activations` keeps all hidden activations in memory, and grows linearly in
+        # space with the number of samples
+        # self._activations = None
+        self._activations = torch.tensor([])
+
+    def get_statistics(self):
+        return (self._activations,)
+
+    def __call__(self, module, x, y):
+        """
+        Forward hook on torch.nn.Module.
+
+        :param module: module
+        :param x: input to module, or tuple of inputs to module
+        :param y: output of module
+        """
+        if not self._tracking:
+            return
+
+        self._activations = self._activations.to(y.device)
+        self._activations = torch.cat((y, self._activations), dim=0)
+
+        # Keep only the last 'num_samples'
+        self._activations = self._activations[:self.num_samples, ...]
```

Comment:
Related to my "post-sigmoid" question, do you intend to track a `torch.nn.Sigmoid` module with this hook? I know it could work for most modules; however, the reason I ask is to make sure you're not intending to track an `ApplyDendrites` module (e.g. `DendriticGate1d`) as these return a namedtuple for which you'd have to access via `y.values`.

Reply:
I don't intend to track a `torch.nn.Sigmoid` module, but it would work for that. In any case, I don't plan to use this hook for dendrites at all; it should only be used where the output is 2D, with shape (batch_size, num_features).